Userland/Libraries/LibGUI/VimEditingEngine.cpp

```diff
@@ -1403,4 +1404,188 @@ void VimEditingEngine::move_to_next_empty_lines_block()
+        KeyCode::Key_Underscore,
+        KeyCode::Key_Plus,
+        KeyCode::Key_Pipe,
+        KeyCode::Key_Tilde,
+        KeyCode::Key_LeftBrace,
+        KeyCode::Key_RightBrace,
+        KeyCode::Key_Colon,
+        KeyCode::Key_DoubleQuote,
+        KeyCode::Key_LessThan,
+        KeyCode::Key_GreaterThan,
+        KeyCode::Key_QuestionMark
+    };
+    return shift_required_keys.contains_slow(key);
+}
+
+String VimKey::to_string() const
+{
+    return String::formatted("key_code: {} string:{} shift:{} ctrl:{}", (int)m_key_code, m_string, m_shift, m_ctrl);
+}
+
+Vector<VimKey> VimCommand::parse_keys(const String& keys)
+{
+    Vector<VimKey> vim_keys {};
+    GenericLexer lexer = GenericLexer(keys);
+    while (!lexer.is_eof()) {
+        char peek = lexer.peek();
+        if (peek == '<') {
+            vim_keys.append(parse_modified_key(lexer));
+        } else {
+            char consumed_char = lexer.consume();
+            StringView key_str = StringView(&consumed_char, 1);
+            KeyCode key_code = string_to_key_code(key_str.to_uppercase_string());
+            bool shift = VimEditingEngine::key_requires_shift(consumed_char, key_code);
+            vim_keys.append(VimKey(key_code, key_str, shift));
+        }
+    }
+    return vim_keys;
+}
+
+VimKey VimCommand::parse_modified_key(GenericLexer& lexer)
+{
+    VERIFY(lexer.consume_specific('<'));
+    bool has_modifier = lexer.peek(1) == '-';
+    char possible_modifier = lexer.peek();
+    if (has_modifier) {
+        VERIFY(possible_modifier == 'S' || possible_modifier == 'C');
+        lexer.consume_specific('-');
+    }
+    StringView key = lexer.consume_until('>');
+    KeyCode key_code = string_to_key_code(key.length() == 1 ? (String)key.to_uppercase_string() : (String)key);
+    bool shift = VimEditingEngine::special_character_requires_shift(key_code);
+    return VimKey(key_code, key, shift || possible_modifier == 'S', possible_modifier == 'C');
+}
+
+void VimCommand::execute() const
+{
+    m_execute_command();
+}
+
+VimMatchType VimCommand::check_match(Vector<VimKey> keys, VimMode vim_mode) const
+{
+    StringView current_vim_mode_char;
+    switch (vim_mode) {
+    case Normal:
+        current_vim_mode_char = "N";
+        break;
+    case Insert:
+        current_vim_mode_char = "I";
+        break;
+    case Visual:
+        current_vim_mode_char = "V";
+        break;
+    }
+
+    if (!m_vim_modes.contains(current_vim_mode_char))
+        return VimMatchType::None;
+
+    if (keys.size() > m_keys.size()) {
+        return VimMatchType::None;
+    }
+
+    for (size_t i = 0; i < keys.size(); i++) {
+        if (m_keys[i] != keys[i]) {
+            return VimMatchType::None;
+        }
+    }
+
+    if (keys.size() < m_keys.size()) {
+        return VimMatchType::Partial;
+    }
+    return VimMatchType::Complete;
+}
+
+String VimCommand::to_string() const
+{
+    StringBuilder sb = StringBuilder();
+    for (const auto& item : m_keys) {
```

Comment:
Please move debugging-related string creation to your `Formatter` specialization.

Also, there is a `Vector<T>` formatter, which will automatically use the formatter of `T`. This might save you some code.